kyo-core/jvm/src/main/scala/kyo/StreamCompression.scala

```diff
@@ -0,0 +1,724 @@
+                    }
+                case DeflateState.PullDeflater(deflater, maybeEmitFn, chunk) =>
+                    local.use: buffer =>
+                        IO(deflater.deflate(buffer, 0, buffer.length, flushMode.value))
+                            .map:
+                                case 0 =>
+                                    Loop.continue(DeflateState.EmitDeflated(deflater, maybeEmitFn, chunk))
+                                case size =>
+                                    IO(chunk.concat(Chunk.fromNoCopy[Byte](java.util.Arrays.copyOf(buffer, size))))
+                                        .map(nextChunk => Loop.continue(DeflateState.PullDeflater(deflater, maybeEmitFn, nextChunk)))
+                case DeflateState.EmitDeflated(deflater, maybeEmitFn, chunk) =>
+                    Emit.andMap(chunk): ack =>
+                        maybeEmitFn match
+                            case Present(emitFn) => Loop.continue(DeflateState.DeflateInput(deflater, emitFn(ack)))
+                            case Absent          => IO(deflater.end()).andThen(Loop.done(Ack.Stop))
+        end emit
+
+        Stream(emit(stream, bufferSize, compressionLevel, strategy, flushMode, noWrap))
+    end deflateStream
+
+    private def gzipStream[Ctx](
+        stream: Stream[Byte, Ctx],
+        bufferSize: Int,
+        compressionLevel: CompressionLevel,
+        strategy: CompressionStrategy,
+        flushMode: FlushMode
+    )(
+        using
+        tag: Tag[Emit[Chunk[Byte]]],
+        _frame: Frame
+    ): Stream[Byte, Resource & IO & Ctx] =
+        enum GZipState derives CanEqual:
+            case Initialize                                                                            extends GZipState
+            case SendHeader(delfater: Deflater, crc32: CRC32)                                          extends GZipState
+            case DeflateInput(delfater: Deflater, crc32: CRC32, emit: Ack < (Emit[Chunk[Byte]] & Ctx)) extends GZipState
+            case PullDeflater(
+                delfater: Deflater,
+                crc32: CRC32,
+                maybeEmitFn: Maybe[Ack => Ack < (Emit[Chunk[Byte]] & Ctx)],
+                chunk: Chunk[Byte]
+            ) extends GZipState
+            case EmitDeflated(
+                delfater: Deflater,
+                crc32: CRC32,
+                maybeEmitFn: Maybe[Ack => Ack < (Emit[Chunk[Byte]] & Ctx)],
+                chunk: Chunk[Byte]
+            )                                                  extends GZipState
+            case SendTrailer(delfater: Deflater, crc32: CRC32) extends GZipState
+        end GZipState
+
+        def emit(
+            stream: Stream[Byte, Ctx],
+            bufferSize: Int,
+            compressionLevel: CompressionLevel,
+            strategy: CompressionStrategy,
+            flushMode: FlushMode
+        ): Ack < (Resource & IO & Emit[Chunk[Byte]] & Ctx) =
+            val local = Local.init(new Array[Byte](bufferSize))
```

Comment:
Why use a local here? Isn't it possible to use the Array directly? This also needs an `IO` suspension since it's allocating a mutable object.

Reply:
`Array[Byte]` here works only as a buffer for `Deflater`/`Inflater` to write data to. Buffer content then will be copied into a `Chunk[Byte]`, so the buffer can be reuse. Using `IO` can lead to unnecessary allocations.

Reply:
Can you use a ChunkBuilder? It should be reusable.